File: tests/algorithms/x/Pascal/dynamic_programming/abbreviation.pas
{$mode objfpc}
program Main;
uses SysUtils;
type BoolArray = array of boolean;
type BoolArrayArray = array of BoolArray;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  n: integer;
  c: string;
  a: string;
  b: boolean;
  s: string;
  ch: string;
function index_of(s: string; ch: string): integer; forward;
function ord_(ch: string): integer; forward;
function chr(n: integer): string; forward;
function to_upper_char(c: string): string; forward;
function is_lower(c: string): boolean; forward;
function abbr(a: string; b: string): boolean; forward;
procedure print_bool(b: boolean); forward;
function index_of(s: string; ch: string): integer;
var
  index_of_i: integer;
begin
  index_of_i := 0;
  while index_of_i < Length(s) do begin
  if s[index_of_i+1] = ch then begin
  exit(index_of_i);
end;
  index_of_i := index_of_i + 1;
end;
  exit(-1);
end;
function ord_(ch: string): integer;
var
  ord__upper: string;
  ord__lower: string;
  ord__idx: integer;
begin
  ord__upper := 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  ord__lower := 'abcdefghijklmnopqrstuvwxyz';
  ord__idx := index_of(ord__upper, ch);
  if ord__idx >= 0 then begin
  exit(65 + ord__idx);
end;
  ord__idx := index_of(ord__lower, ch);
  if ord__idx >= 0 then begin
  exit(97 + ord__idx);
end;
  exit(0);
end;
function chr(n: integer): string;
var
  chr_upper: string;
  chr_lower: string;
begin
  chr_upper := 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  chr_lower := 'abcdefghijklmnopqrstuvwxyz';
  if (n >= 65) and (n < 91) then begin
  exit(copy(chr_upper, n - 65+1, (n - 64 - (n - 65))));
end;
  if (n >= 97) and (n < 123) then begin
  exit(copy(chr_lower, n - 97+1, (n - 96 - (n - 97))));
end;
  exit('?');
end;
function to_upper_char(c: string): string;
var
  to_upper_char_code: integer;
begin
  to_upper_char_code := ord_(c);
  if (to_upper_char_code >= 97) and (to_upper_char_code <= 122) then begin
  exit(chr(to_upper_char_code - 32));
end;
  exit(c);
end;
function is_lower(c: string): boolean;
var
  is_lower_code: integer;
begin
  is_lower_code := ord_(c);
  exit((is_lower_code >= 97) and (is_lower_code <= 122));
end;
function abbr(a: string; b: string): boolean;
var
  abbr_n: integer;
  abbr_m: integer;
  abbr_dp: array of BoolArray;
  abbr_i: integer;
  abbr_row: array of boolean;
  abbr_j: integer;
begin
  abbr_n := Length(a);
  abbr_m := Length(b);
  abbr_dp := [];
  abbr_i := 0;
  while abbr_i <= abbr_n do begin
  abbr_row := [];
  abbr_j := 0;
  while abbr_j <= abbr_m do begin
  abbr_row := concat(abbr_row, [false]);
  abbr_j := abbr_j + 1;
end;
  abbr_dp := concat(abbr_dp, [abbr_row]);
  abbr_i := abbr_i + 1;
end;
  abbr_dp[0][0] := true;
  abbr_i := 0;
  while abbr_i < abbr_n do begin
  abbr_j := 0;
  while abbr_j <= abbr_m do begin
  if abbr_dp[abbr_i][abbr_j] then begin
  if (abbr_j < abbr_m) and (to_upper_char(a[abbr_i+1]) = b[abbr_j+1]) then begin
  abbr_dp[abbr_i + 1][abbr_j + 1] := true;
end;
  if is_lower(a[abbr_i+1]) then begin
  abbr_dp[abbr_i + 1][abbr_j] := true;
end;
end;
  abbr_j := abbr_j + 1;
end;
  abbr_i := abbr_i + 1;
end;
  exit(abbr_dp[abbr_n][abbr_m]);
end;
procedure print_bool(b: boolean);
begin
  if b then begin
  writeln(Ord(true));
end else begin
  writeln(Ord(false));
end;
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  print_bool(abbr('daBcd', 'ABC'));
  print_bool(abbr('dBcd', 'ABC'));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/algorithms/x/Clojure/dynamic_programming/abbreviation.clj
(ns main (:refer-clojure :exclude [index_of ord chr to_upper_char is_lower abbr print_bool]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(defn toi [s]
  (Integer/parseInt (str s)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare index_of ord chr to_upper_char is_lower abbr print_bool)

(def ^:dynamic abbr_dp nil)

(def ^:dynamic abbr_i nil)

(def ^:dynamic abbr_j nil)

(def ^:dynamic abbr_m nil)

(def ^:dynamic abbr_n nil)

(def ^:dynamic abbr_row nil)

(def ^:dynamic chr_lower nil)

(def ^:dynamic chr_upper nil)

(def ^:dynamic index_of_i nil)

(def ^:dynamic is_lower_code nil)

(def ^:dynamic ord_idx nil)

(def ^:dynamic ord_lower nil)

(def ^:dynamic ord_upper nil)

(def ^:dynamic to_upper_char_code nil)

(defn index_of [index_of_s index_of_ch]
  (binding [index_of_i nil] (try (do (set! index_of_i 0) (while (< index_of_i (count index_of_s)) (do (when (= (subs index_of_s index_of_i (+ index_of_i 1)) index_of_ch) (throw (ex-info "return" {:v index_of_i}))) (set! index_of_i (+ index_of_i 1)))) (throw (ex-info "return" {:v (- 1)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn ord [ord_ch]
  (binding [ord_idx nil ord_lower nil ord_upper nil] (try (do (set! ord_upper "ABCDEFGHIJKLMNOPQRSTUVWXYZ") (set! ord_lower "abcdefghijklmnopqrstuvwxyz") (set! ord_idx (index_of ord_upper ord_ch)) (when (>= ord_idx 0) (throw (ex-info "return" {:v (+ 65 ord_idx)}))) (set! ord_idx (index_of ord_lower ord_ch)) (if (>= ord_idx 0) (+ 97 ord_idx) 0)) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn chr [chr_n]
  (binding [chr_lower nil chr_upper nil] (try (do (set! chr_upper "ABCDEFGHIJKLMNOPQRSTUVWXYZ") (set! chr_lower "abcdefghijklmnopqrstuvwxyz") (when (and (>= chr_n 65) (< chr_n 91)) (throw (ex-info "return" {:v (subs chr_upper (- chr_n 65) (min (- chr_n 64) (count chr_upper)))}))) (if (and (>= chr_n 97) (< chr_n 123)) (subs chr_lower (- chr_n 97) (min (- chr_n 96) (count chr_lower))) "?")) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn to_upper_char [to_upper_char_c]
  (binding [to_upper_char_code nil] (try (do (set! to_upper_char_code (ord to_upper_char_c)) (if (and (>= to_upper_char_code 97) (<= to_upper_char_code 122)) (chr (- to_upper_char_code 32)) to_upper_char_c)) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn is_lower [is_lower_c]
  (binding [is_lower_code nil] (try (do (set! is_lower_code (ord is_lower_c)) (throw (ex-info "return" {:v (and (>= is_lower_code 97) (<= is_lower_code 122))}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn abbr [abbr_a abbr_b]
  (binding [abbr_dp nil abbr_i nil abbr_j nil abbr_m nil abbr_n nil abbr_row nil] (try (do (set! abbr_n (count abbr_a)) (set! abbr_m (count abbr_b)) (set! abbr_dp []) (set! abbr_i 0) (while (<= abbr_i abbr_n) (do (set! abbr_row []) (set! abbr_j 0) (while (<= abbr_j abbr_m) (do (set! abbr_row (conj abbr_row false)) (set! abbr_j (+ abbr_j 1)))) (set! abbr_dp (conj abbr_dp abbr_row)) (set! abbr_i (+ abbr_i 1)))) (set! abbr_dp (assoc-in abbr_dp [0 0] true)) (set! abbr_i 0) (while (< abbr_i abbr_n) (do (set! abbr_j 0) (while (<= abbr_j abbr_m) (do (when (nth (nth abbr_dp abbr_i) abbr_j) (do (when (and (< abbr_j abbr_m) (= (to_upper_char (subs abbr_a abbr_i (+ abbr_i 1))) (subs abbr_b abbr_j (+ abbr_j 1)))) (set! abbr_dp (assoc-in abbr_dp [(+ abbr_i 1) (+ abbr_j 1)] true))) (when (is_lower (subs abbr_a abbr_i (+ abbr_i 1))) (set! abbr_dp (assoc-in abbr_dp [(+ abbr_i 1) abbr_j] true))))) (set! abbr_j (+ abbr_j 1)))) (set! abbr_i (+ abbr_i 1)))) (throw (ex-info "return" {:v (nth (nth abbr_dp abbr_n) abbr_m)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn print_bool [print_bool_b]
  (do (if print_bool_b (println true) (println false)) print_bool_b))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (print_bool (abbr "daBcd" "ABC"))
      (print_bool (abbr "dBcd" "ABC"))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)
